cmd/agent/app/reporter/client_metrics_test.go

```diff
@@ -0,0 +1,270 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/uber/jaeger-lib/metrics/metricstest"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest/observer"
+
+	"github.com/jaegertracing/jaeger/cmd/agent/app/testutils"
+	"github.com/jaegertracing/jaeger/thrift-gen/jaeger"
+	"github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
+)
+
+type clientMetricsTest struct {
+	mr   *testutils.InMemoryReporter
+	r    *ClientMetricsReporter
+	logs *observer.ObservedLogs
+	mb   *metricstest.Factory
+}
+
+func (tr *clientMetricsTest) assertLog(t *testing.T, msg, clientUUID string) {
+	logs := tr.logs.FilterMessageSnippet(msg)
+	if clientUUID == "" {
+		assert.Equal(t, 0, logs.Len(), "not expecting log '%s", msg)
+	} else {
+		if assert.Equal(t, 1, logs.Len(), "expecting one log '%s'", msg) {
+			field := logs.All()[0].ContextMap()["client-uuid"]
+			assert.Equal(t, clientUUID, field, "client-uuid should be logged")
+		}
+	}
+}
+
+func testClientMetrics(fn func(tr *clientMetricsTest)) {
+	testClientMetricsWithParams(ClientMetricsReporterParams{}, fn)
+}
+
+func testClientMetricsWithParams(params ClientMetricsReporterParams, fn func(tr *clientMetricsTest)) {
+	r1 := testutils.NewInMemoryReporter()
+	zapCore, logs := observer.New(zap.DebugLevel)
+	mb := metricstest.NewFactory(time.Hour)
+
+	params.Reporter = r1
+	params.Logger = zap.New(zapCore)
+	params.MetricsFactory = mb
+
+	r := WrapWithClientMetrics(params)
+	defer r.Close()
+
+	tr := &clientMetricsTest{
+		mr:   r1,
+		r:    r,
+		logs: logs,
+		mb:   mb,
+	}
+	fn(tr)
+}
+
+func TestClientMetricsReporter_Zipkin(t *testing.T) {
+	testClientMetrics(func(tr *clientMetricsTest) {
+		assert.NoError(t, tr.r.EmitZipkinBatch([]*zipkincore.Span{{}}))
+		assert.Len(t, tr.mr.ZipkinSpans(), 1)
+	})
+}
+
+func TestClientMetricsReporter_Jaeger(t *testing.T) {
+	testClientMetrics(func(tr *clientMetricsTest) {
+		blank := ""
+		clientUUID := "foobar"
+		nPtr := func(v int64) *int64 { return &v }
+		const prefix = "client_stats."
+		tag := func(name, value string) map[string]string { return map[string]string{name: value} }
+
+		tests := []struct {
+			clientUUID  *string
+			seqNo       *int64
+			stats       *jaeger.ClientStats
+			runExpire   bool // invoke expireClientMetrics to update the gauge
+			expLog      string
+			expCounters []metricstest.ExpectedMetric
+			expGauges   []metricstest.ExpectedMetric
+		}{
+			{},
+			{clientUUID: &blank},
+			{clientUUID: &clientUUID},
+			{
+				clientUUID: &clientUUID,
+				seqNo:      nPtr(100),
+				expLog:     clientUUID,
+				runExpire:  true,
+				expCounters: []metricstest.ExpectedMetric{
+					{Name: prefix + "batches_sent", Value: 100},
+				},
+				expGauges: []metricstest.ExpectedMetric{
+					{Name: prefix + "connected_clients", Value: 1},
+				},
+			},
+			{
+				clientUUID: &clientUUID,
+				seqNo:      nPtr(101),
+				expCounters: []metricstest.ExpectedMetric{
+					{Name: prefix + "batches_sent", Value: 101},
+				},
+			},
+			{clientUUID: &clientUUID, seqNo: nPtr(90)},
```

Comment:
Can you add the expected counters/gauges to this test? They should be the same from the previous test (seq nr. 101).

Reply:
done